app/controllers/advanced_messenger_controller.rb

```diff
@@ -1,44 +1,100 @@
       render_403
       return
     end
-    read_by_users = JSON.parse(@journal.read_by_users);
-    if not ["0", "1", "2"].include? params[:read_by_users]
-      Rails.logger.error("#{params[:read_by_users]} is not a valid journal read status for a user")
+    read_by_users = JSON.parse(entity.read_by_users);
+    if not ["0", "1", "2", "3"].include? read_value
+      Rails.logger.error("#{read_value} is not a valid read status for a user")
       render_404
       return
     end
 
-    if read_by_users[User.current.id.to_s] == nil &&  params[:read_by_users] == "2"
+    set_current_user_new_read_collapsed_value(read_by_users, read_value);
+    entity.read_by_users = read_by_users.to_json
+    entity.save
+
+    #setup variables needed in response (in update_journal_read_by_users.js.erb/update_message_read_by_users.js.erb)
+    response_setup.call(entity)
+
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  def set_current_user_new_read_collapsed_value(read_by_users, new_value)
+    if read_by_users[User.current.id.to_s] == nil &&  new_value == "2"
       read_by_users[User.current.id.to_s] = {collapsed: 1}
-    elsif read_by_users[User.current.id.to_s] != nil && read_by_users[User.current.id.to_s]["collapsed"] != nil && params[:read_by_users] == "1"
+    elsif read_by_users[User.current.id.to_s] != nil && read_by_users[User.current.id.to_s]["collapsed"] != nil && new_value == "1"
       read_by_users.delete(User.current.id.to_s)
     else  
-      read_by_users[User.current.id.to_s] = {read: params[:read_by_users].to_i, date: DateTime.now}
+      read_by_users[User.current.id.to_s] = {read: new_value.to_i, date: DateTime.now}
+    end
+  end
+
+  def ignore_all_unread_issues_notes
+    unread_issues_notifications = Journal.where("notes != '' AND notes IS NOT NULL AND read_by_users ILIKE ?", '%"' + User.current.id.to_s + '":{"read":0%');
+    ignore_all_unread_entities(unread_issues_notifications)
+  end
+
+  def ignore_all_unread_issues_notes_for_an_issue
+    unread_issue_notifications = Journal.where("notes != '' AND notes IS NOT NULL AND journalized_id = ? AND read_by_users ILIKE ?", params[:issue_id], '%"' + User.current.id.to_s + '":{"read":0%');
+    ignore_all_unread_entities(unread_issue_notifications)
+  end
+
+  def ignore_all_unread_forum_messages
+    unread_forum_messages = Message.where("read_by_users ILIKE ?", '%"' + User.current.id.to_s + '":{"read":0%');
+    ignore_all_unread_entities(unread_forum_messages)
+  end
+
+  def ignore_all_unread_forum_messages_for_a_topic
+    unread_messages_for_topic = Message.where("(id = ? OR parent_id = ?) AND read_by_users ILIKE ?", params[:topic_id], params[:topic_id], '%"' + User.current.id.to_s + '":{"read":0%');
+    ignore_all_unread_entities(unread_messages_for_topic)
+  end
+
+  def ignore_all_unread_entities(unread_entities)
+    unread_entities.each_with_index do |unread_entity, index|
+      read_by_users = JSON.parse(unread_entity.read_by_users) 
```

Comment:
I think we should also check if the note/forum message is visible by the user, right? (as we did in the rest of implementation).

I'm wondering, can a note has a `{"red":0}` in db and not be visible for the current user?

Reply:
A note is for sure visible for a user when the  ```{"red":0}``` is initialy set, but it can become invisible afterwards. I.e. the note becomes private by editing it and the user doesn't have the right to view private notes

So I changed this method to filter and ignore only the visible entities (journal/message) for the current user